### ISSUE TITLE
twisted.conch.ssh.filetransfer.FileTransferServer leaks file descriptors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3213]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3213|
|Type|enhancement|
|Created|2008-04-29 12:39:35Z|
|Branch|https://github.com/twisted/twisted/tree/filetransfer-leak-3213|

If a remote client opens a file but doesn't close it cleanly (for example, if the connection dies), the file remained open (it uses os.open, so garbage collect doesn't close it). All the opened files should be closed at connectionLost.

This is (somehow) the problem spotted in [#1789](https://github.com/twisted/twisted/issues/1789).

<details><summary>Searchable metadata</summary>

```
trac-id__3213 3213
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_filetransfer_leak_3213 branches/filetransfer-leak-3213
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1209472775000000 1209472775000000
changetime__1211623477000000 1211623477000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23458]) Branching to 'filetransfer-leak-3213'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

This is done in r23459. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

Looks pretty good.  I'd be happiest if we had some kind of general fs-api mock that could easily be installed and removed and about the state of which assertions could easily be made (eg, assert that it has no open files), but it'd be dumb to block this ticket on that kind of thing.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set status to `closed`|
|-|-|

(In [23695]) Merge filetransfer-leak-3213

Author: therve
Reviewer: exarkun
Fixes [#3213](https://github.com/twisted/twisted/issues/3213)
Fixes [#1789](https://github.com/twisted/twisted/issues/1789)

Implement connectionLost on twisted.conch.filetransfer.FileTransferServer to
close all opened files and directories: previously, if a connection died
unexpectedly, all opened files were leaked.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
